[quant] Input Weight Equalization - prepare modifications

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59953 [quant] Equalization Observer modifications
* **#59747 [quant] Input Weight Equalization - prepare modifications**
* #59739 [quant] EqualizationQConfig to distinguish input/output activations

Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpQuantObs -> InpEqObs -> linear1 -> OutQuantObs -> x1`
  `w -> WeightQuantObs -> WeightEqObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpQuantObs1 -> InpEqObs1 -> linear1 -> OutQuantObs1 -> InpEqObs2 -> linear2 -> OutQuantObs2 -> x1`
  `w1 -> WeightQuantObs1 -> WeightEqObs1 -> linear1`, `w2 -> WeightQuantObs2 -> WeightEqObs2 -> linear2

Test Plan: `python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare`

Original model with one `nn.Linear` layer
```
LinearModule(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %linear : [#users=1] = call_module[target=linear](args = (%x_activation_post_process_0_equalization_process_0,), kwargs = {})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    return linear_activation_post_process_0
```
--------------------------------------

Original model with two connected functional linear layers
```
FunctionalLinearModule(
  (linear1): Linear()
  (linear2): Linear()
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x,), kwargs = {})
    %x_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x_activation_post_process_0,), kwargs = {})
    %linear1_w : [#users=1] = get_attr[target=linear1.w]
    %linear1_w_equalization_process_0 : [#users=1] = call_module[target=linear1_w_equalization_process_0](args = (%linear1_w,), kwargs = {})
    %linear1_w_activation_post_process_0 : [#users=1] = call_module[target=linear1_w_activation_post_process_00](args = (%linear1_w_equalization_process_0,), kwargs = {})
    %linear1_b : [#users=1] = get_attr[target=linear1.b]
    %linear : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%x_activation_post_process_0_equalization_process_0, %linear1_w_activation_post_process_0), kwargs = {bias: %linear1_b})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    %linear_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0_equalization_process_0](args = (%linear_activation_post_process_0,), kwargs = {})
    %linear2_w : [#users=1] = get_attr[target=linear2.w]
    %linear2_w_equalization_process_0 : [#users=1] = call_module[target=linear2_w_equalization_process_0](args = (%linear2_w,), kwargs = {})
    %linear2_w_activation_post_process_0 : [#users=1] = call_module[target=linear2_w_activation_post_process_00](args = (%linear2_w_equalization_process_0,), kwargs = {})
    %linear2_b : [#users=1] = get_attr[target=linear2.b]
    %linear_1 : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%linear_activation_post_process_0_equalization_process_0, %linear2_w_activation_post_process_0), kwargs = {bias: %linear2_b})
    %linear_1_activation_post_process_0 : [#users=1] = call_module[target=linear_1_activation_post_process_0](args = (%linear_1,), kwargs = {})
    return linear_1_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135316](https://our.internmc.facebook.com/intern/diff/D29135316)

angelayi committed:
[quant] Input Weight Equalization - prepare modifications

Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -273,8 +275,12 @@ def maybe_insert_input_observer_for_arg_or_kwarg(
+
         act_post_process_ctr = qconfig.weight if is_weight else \
             qconfig.activation
+        if isinstance(qconfig, EqualizationQConfig) and not is_weight:
```
I'd recommend rewriting this so the `act_post_process_ctr` value gets set exactly once and is not modified, to minimize the chance of future bugs.  For example,

```
if is_weight:
  ...
elif isinstance(qconfig, EqualizationQConfig):
  ...
else:
  ...
```

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -458,7 +464,23 @@ def maybe_insert_output_observer_for_node(
                         matched_pattern,
                         act_post_process_ctr)
+
+            # Find the next node in the graph
+            next_node = node.next
```
this seems unintuitive, is there a design document which explains what we are trying to do?

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Modifies prepare_fx for input-weight equalization. Ensures that
the output observer that is inserted is set as the output_obs value
specified by the user.

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented on test/quantization/fx/test_equalize_fx.py:
```diff
@@ -110,3 +122,108 @@ def test_input_weight_observer(self, input_qdtype, input_qscheme, weight_qdtype,
+        }
+
+        functionalLinearFP32_node_occurrence = {
+            ns.call_module(_InputEqualizationObserver): 4,
```
Why do we expect 4 InputEqualizationObservers?

supriyar commented on torch/quantization/quantize_fx.py:
```diff
@@ -141,10 +141,12 @@ def create_node(self, kind : str, target : Target,
 
 def _prepare_fx(model: torch.nn.Module, qconfig_dict: Any,
                 prepare_custom_config_dict: Dict[str, Any] = None,
+                equalization_qconfig_dict: Dict[str, Any] = None,
```
Should we add this to the end with an `_` prefix since this is still in development? @vkuzo what are your thoughts?

supriyar replied:
yeah originally we thought of keeping it in the `qconfig_dict` but decided to separate it out to make it more explicit since equalization is a different step compared to quantization.

supriyar replied:
We can move the equalization + quantization observers into EqualizationQConfig. So we'd have the following fields - 
`activation`, `weight`, `activation_eq` and `weight_eq`. @vkuzo would this make the code in `prepare` mode complex since we have to check for these fields for all nodes now?

vkuzo replied:
> yeah originally we thought of keeping it in the qconfig_dict but decided to separate it out to make it more explicit since equalization is a different step compared to quantization.

makes sense, thanks for explaining, never mind

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -812,7 +840,8 @@ def insert_observers_for_model(
                         node, qconfig, model, modules, graph,
                         node_name_to_target_dtype,
-                        qhandler, prepare_custom_config_dict)
+                        qhandler, prepare_custom_config_dict,
+                        equalization_qconfig)
```
the logic for inserting observers for equalization seems a lot simpler (at least at the start).  Can we consider keeping this implenetation fully separate?  For example, we could have:

```
for node in nodes_before_observation:
  ...
  # insert quantization input observers, if needed
  # insert equalization input observers, if needed
  # insert quantization output observers, if needed
```

That way the code for equalization is separate in its own functions, instead of being mixed with the quantization code.

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -412,6 +419,44 @@ def maybe_insert_input_observers_for_node(
     node.args = tuple(new_args)
     node.kwargs = new_kwargs  # type: ignore[assignment]
 
+def maybe_insert_input_equalization_observers_for_node(
```
is equalization only supported for `nn.Linear` and `F.linear` nodes, or is it going to be supported more generically for more types of nodes?

If it's only linears, might make sense to write this function specifically for linears instead of making it generic.

angelayi replied:
I think in the future we hope to support equalization for more types of nodes (ex. conv) but I think as of right now, it only supports linear nodes.

vkuzo replied:
in that case, it would be good to have something like

```
# in equalization specific config file
def node_supports_equalization(node: Node) -> bool

# in prepare.py, in function which inserts equalization observers
if not node_supports_equalization(node):
    return
... continue with inserting equalization observers...
```

angelayi replied:
If I'm understanding this correctly, is the purpose to remove the passing of the `equalization_qconfig_dict` and setting it as a global variable in `_equalize.py` so that we can abstract the checking of if the node supports equalization?

Right now, we check to see if the node supports equalization by seeing if it exists in `equalization_qconfig_map`, so does that mean that instead of doing this, we would move all logic to `node_supports_equalization`?

vkuzo replied:
I thought `equalization_qconfig_map` just checks what the user asked for.  Does it actually do any checking whether what the user asked for is supported?

angelayi replied:
Sorry for the late response!! I think I understand what you mean. For now, we would checking for if the node is an `nn.Linear` or `F.linear`?

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -160,7 +166,13 @@ def insert_observer(
+    if isinstance(observer, _InputEqualizationObserver) or isinstance(observer, _WeightEqualizationObserver):
+        prefix = node.name + '_equalization_process_'
+    else:
+        if 'equalization' in node.name:
+            prefix = node.name.replace('equalization', 'activation_post')
```
what's the context on this?  Could you give an example of when this code path is taken?

It would be good to avoid replacing parts of strings, it's really convenient to have the entire prefix in the code so that it is easy to grep for.

angelayi replied:
In the original code (without the equalization observers), the graph after prepare for `x -> InputQuantObs` would look something like: `x -> x_activation_post_process_0`. The naming scheme is just taking the name of the previous node (`x`) and adding `_activation_post_process_`.

After the input equalization observer is be inserted before the input quantization observer, the graph after prepare would something like `x -> InputEqObs -> InputQuantObs` and would now be labeled as `x -> x_equalization_process_0 -> x_equalization_process_0_activation_post_process_0`. 

Since this naming is slightly inaccurate as the `InputQuantObs` should be `x`'s activation post process, I tried to get rid of the additional `x_equalization_process_0` in the third term by removing the entire `equalization` part in the name and replacing it with `activation`.

vkuzo replied:
makes sense, thanks for explaining.  Might be good to just keep the naming simple (`x_equalization_process_0_activation_post_process_0` sounds ok), it would be easier to debug that way.

that said, are we sure this order is what we want?  I would have expected the quantization observer to come first, and the equalization observer to come second.  For example, if we have `linear1 -> linear2`, we would have `eq_obs1 -> linear1 -> obs1` inserted when processing linear1, and `eq_obs2 -> linear2 -> obs2` inserted when processing linear2.  So, the order between `linear1` and `linear2` is quantization observer first, and equalization observer second.

supriyar replied:
I think the motivation for this was that the equalization observer gets replaced by a mul and the quantization observer gets replaced by a quantize_per_tensor (for fp32 input) and the expected graph after convert is `x -> mul -> quantize_per_tensor -> quantized_lienar -> dequant`, so the prepared graph is `x -> eq_obs -> in_obs -> linear -> out_obs`

For the 2 linear case, I think the order would be `x -> eq_obs -> inp_obs -> linear1 -> out_obs -> eq_obs2 -> linear2 -> out_obs2` (since linear1 out_obs gets inserted first). 
I think this can be changed to have the eq_obs appear after the quantization observer for the first linear (with fp32 input), it shouldn't affect the functionality. @angelayi lmk what you think

angelayi replied:
> that said, are we sure this order is what we want? 

This might be a little confusing, but here are the cases:
- For the input layer, it should look like: `x -> eq_obs -> in_obs -> linear`
- This is this same as when there is a non-quantized operation (assume `add` is a `fp32` op): `add -> eq_obs -> in_obs -> linear`
- For connected layers, it is like you said, `linear1 -> out_obs1 -> eq_obs2 -> linear2 -> out_obs2`. However, in this case we do not need to scale the next input quantization observer so the specific location of the equalization observer is not really necessary. Instead, we can scale `linear1`'s weight value by the next equalization scale (I have a PR coming for this soon).

vkuzo replied:
yes, unfortunately in this case (at least in how I think it would be the cleanest to build) the order of observers in the graph is not going to match the order of nodes after convert.  It would be

```
# after prepare
x0 -> qobs0 -> eqobs1 -> linear1

# after convert
x0 -> mul(from eqobs1) -> quantize_per_tensor(from qobs0) -> qlinear1
```

I believe this is better because it allows us to keep the invariant that in prepare, we do not need to modify the graph before the current node in order to insert observers for the current node.

angelayi replied:
> For the 2 linear case, I think the order would be x -> eq_obs -> inp_obs -> linear1 -> out_obs -> eq_obs2 -> linear2 -> out_obs2 (since linear1 out_obs gets inserted first).
> I think this can be changed to have the eq_obs appear after the quantization observer for the first linear (with fp32 input), it shouldn't affect the functionality.

Yes, for the 2 connected layers case, it is currently: `linear1 -> out_obs1 -> eq_obs2 -> linear2 -> out_obs2`. However, in this case we do not need to scale the next input quantization observer so the specific location of the equalization observer (whether it is before or after `out_obs1`) is not really necessary. Instead, we can scale `linear1`'s weight value by the next equalization scale (I have a PR coming for this soon).

I'm not sure how to exactly change the existing code to insert `eq_obs2` before `out_obs1`, but I agree that it shouldn't affect the functionality. The reason why I'm not totally sure how to change the existing code is that I think right now the observers are inserted in the following manner:
1. Initially we have `x -> linear1 -> linear2`
2. We look `linear1` and insert the equalization observer, quantization input observer, and quantization output observer: `x -> eq_obs1 -> in_obs1 -> linear1 -> out_obs1 -> linear2`
3. We look at `linear2` and insert the equalization observer: `x -> eq_obs1 -> in_obs1 -> linear1 -> out_obs1 -> eq_obs2 -> linear2`
4. (still looking at `linear2`), we cannot insert the input equalization observer because it has the same type as the `out_obs1` (we do not want to insert duplicate observers).
If we want to insert the `eq_obs2` before `out_obs1` then we would need to move some of the nodes around (if that makes sense).

vkuzo commented on torch/quantization/fx/prepare.py:
```diff
@@ -417,6 +431,44 @@ def maybe_insert_input_observers_for_node(
+    new_args = []
+    for arg in node.args:
+        if not isinstance(arg, Node) or node_arg_is_bias(node, arg):
+            continue
```
for this case, should we assign `arg` to `new_args` instead of discarding?

angelayi replied:
Yes! I will update that!

vkuzo reviewed (commented):
looks good! I had a couple of minor questions inline.

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: `python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare`

Original model with one `nn.Linear` layer
```
LinearModule(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear : [#users=1] = call_module[target=linear](args = (%x_activation_post_process_0,), kwargs = {})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    return linear_activation_post_process_0
```
--------------------------------------

Original model with two connected functional linear layers
```
FunctionalLinearModule(
  (linear1): Linear()
  (linear2): Linear()
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear1_w : [#users=1] = get_attr[target=linear1.w]
    %linear1_w_equalization_process_0 : [#users=1] = call_module[target=linear1_w_equalization_process_0](args = (%linear1_w,), kwargs = {})
    %linear1_w_activation_post_process_0 : [#users=1] = call_module[target=linear1_w_activation_post_process_00](args = (%linear1_w_equalization_process_0,), kwargs = {})
    %linear1_b : [#users=1] = get_attr[target=linear1.b]
    %linear : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%x_activation_post_process_0, %linear1_w_activation_post_process_0), kwargs = {bias: %linear1_b})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    %linear_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0_equalization_process_0](args = (%linear_activation_post_process_0,), kwargs = {})
    %linear2_w : [#users=1] = get_attr[target=linear2.w]
    %linear2_w_equalization_process_0 : [#users=1] = call_module[target=linear2_w_equalization_process_0](args = (%linear2_w,), kwargs = {})
    %linear2_w_activation_post_process_0 : [#users=1] = call_module[target=linear2_w_activation_post_process_00](args = (%linear2_w_equalization_process_0,), kwargs = {})
    %linear2_b : [#users=1] = get_attr[target=linear2.b]
    %linear_1 : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%linear_activation_post_process_0_equalization_process_0, %linear2_w_activation_post_process_0), kwargs = {bias: %linear2_b})
    %linear_1_activation_post_process_0 : [#users=1] = call_module[target=linear_1_activation_post_process_0](args = (%linear_1,), kwargs = {})
    return linear_1_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: `python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare`

Original model with one `nn.Linear` layer
```
LinearModule(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear : [#users=1] = call_module[target=linear](args = (%x_activation_post_process_0,), kwargs = {})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    return linear_activation_post_process_0
```
--------------------------------------

Original model with two connected functional linear layers
```
FunctionalLinearModule(
  (linear1): Linear()
  (linear2): Linear()
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear1_w : [#users=1] = get_attr[target=linear1.w]
    %linear1_w_equalization_process_0 : [#users=1] = call_module[target=linear1_w_equalization_process_0](args = (%linear1_w,), kwargs = {})
    %linear1_w_activation_post_process_0 : [#users=1] = call_module[target=linear1_w_activation_post_process_00](args = (%linear1_w_equalization_process_0,), kwargs = {})
    %linear1_b : [#users=1] = get_attr[target=linear1.b]
    %linear : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%x_activation_post_process_0, %linear1_w_activation_post_process_0), kwargs = {bias: %linear1_b})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    %linear_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0_equalization_process_0](args = (%linear_activation_post_process_0,), kwargs = {})
    %linear2_w : [#users=1] = get_attr[target=linear2.w]
    %linear2_w_equalization_process_0 : [#users=1] = call_module[target=linear2_w_equalization_process_0](args = (%linear2_w,), kwargs = {})
    %linear2_w_activation_post_process_0 : [#users=1] = call_module[target=linear2_w_activation_post_process_00](args = (%linear2_w_equalization_process_0,), kwargs = {})
    %linear2_b : [#users=1] = get_attr[target=linear2.b]
    %linear_1 : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%linear_activation_post_process_0_equalization_process_0, %linear2_w_activation_post_process_0), kwargs = {bias: %linear2_b})
    %linear_1_activation_post_process_0 : [#users=1] = call_module[target=linear_1_activation_post_process_0](args = (%linear_1,), kwargs = {})
    return linear_1_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135316).

supriyar commented on torch/quantization/fx/prepare.py:
```diff
@@ -417,6 +431,45 @@ def maybe_insert_input_observers_for_node(
+        new_eq_obs_node = insert_observer(
+            arg, new_eq_obs_mod, model, modules, graph)
+
+        # set the type, so the next node can read it
+        node_name_to_target_dtype[new_eq_obs_node.name] = node_name_to_target_dtype[arg.name]
```
Do we need to set this type for equalization observers? My understanding is that we want the quantization observers to be inserted irrespective of the equalization observers so they shouldn't really be checking the dtype of the eq observer. @vkuzo lmk if this sounds right

angelayi replied:
I tried to set it up so that the quantization observers are inserted irrespective of the equalization observers by setting the dtype of the equalization observers to be the same dtype as the input. In this case, the quantization observers will be essentially checking the input's dtype instead of the equalization observer's dtype.

vkuzo replied:
I think it makes sense to set the dtype, we should know the dtype for every node of the graph.

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: `python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare`

Original model with one `nn.Linear` layer
```
LinearModule(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear : [#users=1] = call_module[target=linear](args = (%x_activation_post_process_0,), kwargs = {})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    return linear_activation_post_process_0
```
--------------------------------------

Original model with two connected functional linear layers
```
FunctionalLinearModule(
  (linear1): Linear()
  (linear2): Linear()
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear1_w : [#users=1] = get_attr[target=linear1.w]
    %linear1_w_equalization_process_0 : [#users=1] = call_module[target=linear1_w_equalization_process_0](args = (%linear1_w,), kwargs = {})
    %linear1_w_activation_post_process_0 : [#users=1] = call_module[target=linear1_w_activation_post_process_00](args = (%linear1_w_equalization_process_0,), kwargs = {})
    %linear1_b : [#users=1] = get_attr[target=linear1.b]
    %linear : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%x_activation_post_process_0, %linear1_w_activation_post_process_0), kwargs = {bias: %linear1_b})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    %linear_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0_equalization_process_0](args = (%linear_activation_post_process_0,), kwargs = {})
    %linear2_w : [#users=1] = get_attr[target=linear2.w]
    %linear2_w_equalization_process_0 : [#users=1] = call_module[target=linear2_w_equalization_process_0](args = (%linear2_w,), kwargs = {})
    %linear2_w_activation_post_process_0 : [#users=1] = call_module[target=linear2_w_activation_post_process_00](args = (%linear2_w_equalization_process_0,), kwargs = {})
    %linear2_b : [#users=1] = get_attr[target=linear2.b]
    %linear_1 : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%linear_activation_post_process_0_equalization_process_0, %linear2_w_activation_post_process_0), kwargs = {bias: %linear2_b})
    %linear_1_activation_post_process_0 : [#users=1] = call_module[target=linear_1_activation_post_process_0](args = (%linear_1,), kwargs = {})
    return linear_1_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135316](https://our.internmc.facebook.com/intern/diff/D29135316)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/prepare.py:
```diff
@@ -160,7 +166,10 @@ def insert_observer(
         observer.to(model_device)
     # add observer module as attribute
-    prefix = node.name + '_activation_post_process_'
+    if isinstance(observer, _InputEqualizationObserver) or isinstance(observer, _WeightEqualizationObserver):
```
nit: maybe put this check in a function like https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantize.py#L23

jerryzh168 commented on torch/quantization/fx/prepare.py:
```diff
@@ -701,6 +751,7 @@ def insert_observers_for_model(
     qconfig_map: Dict[str, QConfigAny],
     graph: Graph,
     prepare_custom_config_dict: Dict[str, Any],
+    equalization_config_map: Dict[str, Any],
```
why do we need an equalization_config_map? can we include this in qconfig_map?

jerryzh168 commented on torch/quantization/fx/prepare.py:
```diff
@@ -910,19 +967,22 @@ def save_state(
         qconfig_dict: Any,
         node_name_to_scope: Dict[str, Tuple[str, type]],
         prepare_custom_config_dict: Optional[Dict[str, Any]],
+        equalization_qconfig_dict: Optional[Dict[str, Any]],
```
why do we need an equalization_config_map? can we include this in qconfig_map?

jerryzh168 replied:
I see, sounds good

jerryzh168 reviewed (approved):
Looks good

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: `python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare`

Original model with one `nn.Linear` layer
```
LinearModule(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear : [#users=1] = call_module[target=linear](args = (%x_activation_post_process_0,), kwargs = {})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    return linear_activation_post_process_0
```
--------------------------------------

Original model with two connected functional linear layers
```
FunctionalLinearModule(
  (linear1): Linear()
  (linear2): Linear()
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear1_w : [#users=1] = get_attr[target=linear1.w]
    %linear1_w_equalization_process_0 : [#users=1] = call_module[target=linear1_w_equalization_process_0](args = (%linear1_w,), kwargs = {})
    %linear1_w_activation_post_process_0 : [#users=1] = call_module[target=linear1_w_activation_post_process_00](args = (%linear1_w_equalization_process_0,), kwargs = {})
    %linear1_b : [#users=1] = get_attr[target=linear1.b]
    %linear : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%x_activation_post_process_0, %linear1_w_activation_post_process_0), kwargs = {bias: %linear1_b})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    %linear_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0_equalization_process_0](args = (%linear_activation_post_process_0,), kwargs = {})
    %linear2_w : [#users=1] = get_attr[target=linear2.w]
    %linear2_w_equalization_process_0 : [#users=1] = call_module[target=linear2_w_equalization_process_0](args = (%linear2_w,), kwargs = {})
    %linear2_w_activation_post_process_0 : [#users=1] = call_module[target=linear2_w_activation_post_process_00](args = (%linear2_w_equalization_process_0,), kwargs = {})
    %linear2_b : [#users=1] = get_attr[target=linear2.b]
    %linear_1 : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%linear_activation_post_process_0_equalization_process_0, %linear2_w_activation_post_process_0), kwargs = {bias: %linear2_b})
    %linear_1_activation_post_process_0 : [#users=1] = call_module[target=linear_1_activation_post_process_0](args = (%linear_1,), kwargs = {})
    return linear_1_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135316](https://our.internmc.facebook.com/intern/diff/D29135316)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135316).

angelayi committed:
Update on "[quant] Input Weight Equalization - prepare modifications"


Summary: Summary: Modifies prepare_fx for input-weight equalization. If a current
node is being equalized (there exists a EqualizationQConfig), then the
EqualizationObserver will be inserted before its quantization observer.

For a singular linear layer, the general flow looks like:
Original graph: `x0 -> linear -> x1`, `w -> linear`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> x1`
  `w -> WeightEqObs -> MinMaxObs -> linear1`

For two connected linear layers, the general flow looks like:
Original graph: `x0 -> linear1 -> linear2 -> x1`,
  `w1 -> linear1`, `w2 -> linear2`
After prepare: `x0 -> InpEqObs -> MinMaxObs -> linear1 -> MinMaxObs -> InpEqObs -> linear2 -> MinMaxObs -> x1`
  `w1 -> WeightEqObs -> MinMaxObs -> linear1`, `w2 -> WeightEqObs -> MinMaxObs -> linear2

Test Plan: `python test/test_quantization.py
TestEqualizeFx.test_input_equalization_prepare`

Original model with one `nn.Linear` layer
```
LinearModule(
  (linear): Linear(in_features=1, out_features=1, bias=True)
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear : [#users=1] = call_module[target=linear](args = (%x_activation_post_process_0,), kwargs = {})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    return linear_activation_post_process_0
```
--------------------------------------

Original model with two connected functional linear layers
```
FunctionalLinearModule(
  (linear1): Linear()
  (linear2): Linear()
)
```

Graph after `prepare_fx`:
```
graph():
    %x : [#users=1] = placeholder[target=x]
    %x_equalization_process_0 : [#users=1] = call_module[target=x_equalization_process_0](args = (%x,), kwargs = {})
    %x_activation_post_process_0 : [#users=1] = call_module[target=x_activation_post_process_00](args = (%x_equalization_process_0,), kwargs = {})
    %linear1_w : [#users=1] = get_attr[target=linear1.w]
    %linear1_w_equalization_process_0 : [#users=1] = call_module[target=linear1_w_equalization_process_0](args = (%linear1_w,), kwargs = {})
    %linear1_w_activation_post_process_0 : [#users=1] = call_module[target=linear1_w_activation_post_process_00](args = (%linear1_w_equalization_process_0,), kwargs = {})
    %linear1_b : [#users=1] = get_attr[target=linear1.b]
    %linear : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%x_activation_post_process_0, %linear1_w_activation_post_process_0), kwargs = {bias: %linear1_b})
    %linear_activation_post_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0](args = (%linear,), kwargs = {})
    %linear_activation_post_process_0_equalization_process_0 : [#users=1] = call_module[target=linear_activation_post_process_0_equalization_process_0](args = (%linear_activation_post_process_0,), kwargs = {})
    %linear2_w : [#users=1] = get_attr[target=linear2.w]
    %linear2_w_equalization_process_0 : [#users=1] = call_module[target=linear2_w_equalization_process_0](args = (%linear2_w,), kwargs = {})
    %linear2_w_activation_post_process_0 : [#users=1] = call_module[target=linear2_w_activation_post_process_00](args = (%linear2_w_equalization_process_0,), kwargs = {})
    %linear2_b : [#users=1] = get_attr[target=linear2.b]
    %linear_1 : [#users=1] = call_function[target=torch.nn.functional.linear](args = (%linear_activation_post_process_0_equalization_process_0, %linear2_w_activation_post_process_0), kwargs = {bias: %linear2_b})
    %linear_1_activation_post_process_0 : [#users=1] = call_module[target=linear_1_activation_post_process_0](args = (%linear_1,), kwargs = {})
    return linear_1_activation_post_process_0
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29135316](https://our.internmc.facebook.com/intern/diff/D29135316)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29135316).
